ports/nrf/internal_flash.c

```diff
@@ -67,101 +84,63 @@ uint32_t internal_flash_get_block_count(void) {
     return ((uint32_t) __fatfs_flash_length) / FILESYSTEM_BLOCK_SIZE ;
 }
 
+// TODO support flashing with SD enabled
 void internal_flash_flush(void) {
-}
+    if (_fl_pg_addr == NO_CACHE) return;
 
-void flash_flush(void) {
-    internal_flash_flush();
-}
+    // Skip if data is the same
+    if (memcmp(_fl_cache, (void *)_fl_pg_addr, FL_PAGE_SZ) != 0) {
+//        _is_flashing = true;
+        nrf_nvmc_page_erase(_fl_pg_addr);
+        nrf_nvmc_write_words(_fl_pg_addr, (uint32_t *)_fl_cache, FL_PAGE_SZ / sizeof(uint32_t));
+    }
 
-static uint32_t convert_block_to_flash_addr(uint32_t block) {
-  return ((uint32_t)__fatfs_flash_start_addr) + block * FILESYSTEM_BLOCK_SIZE;
+    _fl_pg_addr = NO_CACHE;
 }
 
-bool internal_flash_write_block(const uint8_t *src, uint32_t block) {
-  uint8_t sd_en = 0;
-
-#ifdef MICROPY_HW_LED_MSC
-  port_pin_set_output_level(MICROPY_HW_LED_MSC, true);
-#endif
-  temp_status_color(ACTIVE_WRITE);
-  // non-MBR block, copy to cache
-
-  uint32_t dest = convert_block_to_flash_addr(block);
-
-  uint32_t pagenum = dest / FLASH_PAGE_SIZE;
-  uint32_t* flash_align = (uint32_t*) (pagenum*FLASH_PAGE_SIZE);
-
-  // Read back current page to update only 512 portion
-  __ALIGN(4) uint8_t buf[FLASH_PAGE_SIZE];
-  memcpy(buf, flash_align, FLASH_PAGE_SIZE);
-  memcpy(buf + (dest%FLASH_PAGE_SIZE), src, FILESYSTEM_BLOCK_SIZE);
-
-#ifdef BLUETOOTH_SD
-  (void) sd_softdevice_is_enabled(&sd_en);
-
-  if (sd_en) {
-    if (NRF_SUCCESS != sd_flash_page_erase(pagenum)) {
-      return false;
-    }
-
-    if (NRF_SUCCESS != sd_flash_write(flash_align, (uint32_t*) buf, FLASH_PAGE_SIZE / sizeof(uint32_t))) {
-      return false;
-    }
-  }
-#endif
+mp_uint_t internal_flash_read_blocks(uint8_t *dest, uint32_t block, uint32_t num_blocks) {
+    uint32_t src = lba2addr(block);
+    memcpy(dest, (uint8_t*) src, FILESYSTEM_BLOCK_SIZE*num_blocks);
```

Comment:
add spaces around * operator

Reply:
I actually prefer the other way around, but we indeed should stay consistency as much as possibl